### PR TITLE
Allow users to join existing lists

### LINKED ISSUE #5: 5. As a user, I want to join an existing shopping list so I can share a shopping list with another person
## Summary
We want to allow users to join existing shopping lists so they can manage them with friends or family. This would take the following steps:
1. User enters an existing token into a form
2. The token is saved to `localStorage`
3. The user is taken to the `List` view and shown the existing list.

## Acceptance criteria
If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Notes
- Be sure to use the `setListToken` function to save the token that the user enters!
- You will know your `<label>` is correctly implemented if you click on the label and keyboard focus moves to the related input. Refer to [the MDN docs on `<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) for more info.
- Users will be able to submit the form with the `Enter` key if you listen for the right JavaScript event on the right element! If you're adding an `onKeyDown` listener anywhere, you might be overcomplicating things.
- You can assume that users will only be sharing a token if the list has actual items on it. So, let's say I create a New Shopping List, but never add any items to it, and haven't saved that particular token to Firestore, we can assume that's a list that doesn't exist.
- If you would like, you can save the token to Firestore immediately when creating a list, and then a user would be able to share an empty list (see issue 3).

### PR DESCRIPTION
**Pull request by rdpfeifle**
## Description

Allow users to join existing shopping lists so they can manage them with friends or family by taking the following steps:

1. User enters an existing token into a form
2. The token is saved to both localStorage and Firestore
3. The user is redirected to the List view, where they are shown the existing list or a message indicating that the list is empty.
4. If the list does not exist, a message is shown to the user.

## Improvements
- We created a function that allows a newly created list to be saved to Firestore, even if it has no items. In this way, we enable users to share an empty list.

## Accessibility
- We employed the useRef hook to create a pointer to the token input field to ensure it remains in focus after being cleared.

## Related Issue

closes #5 

## Acceptance Criteria

If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓   | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Files modified
- firebase.js
- Home.jsx
- List.jsx
- Button.jsx

### Before

Homepage before

### After

Join List

## Testing Steps / QA Criteria

- To pull our branch locally, run `git pull origin rp-cm-joinExistingList` 
- Next, run `git checkout rp-cm-joinExistingList`
- Launch the app with `npm start`

### Scenario 1: Checking an **existing** token
- Use the token `ann pure penh` to view an existing list.

### Scenario 2: Checking an **invalid** token
- Try entering a random combination or simply submit with an empty input to test an invalid token.

### Scenario 3: Creating a new list
- Click the button to create a new list
- Check Firestore to confirm that the new list has been added.

### Notes
- If you click the button to clear the input field, you will notice that it will refocus on the input.

### CODE REVIEW

File: src/views/Home.jsx
Comment (AslanTaheri) [inline src/views/Home.jsx:1]: Nice job on all the accessibility considerations and the useRef for focusing back on the input field!

**Review by AslanTaheri — APPROVED** [review]

**Review by shangguanwang — APPROVED** [review]
Very thoughtful approach, and thank you for modifying the Button components input to incorporate the new ariaLabel props we've added

File: src/api/firebase.js
Comment (AslanTaheri) [inline src/api/firebase.js:1]: Taking a second look at this, I am wondering if adding an empty doc would cause issues further down the road, especially since we are having conditional rendering in List.jsx that checks these records to verify whether a list is empty!

Comment (rdpfeifle) [reply src/api/firebase.js:1]: Thank you, @AslanTaheri! You've raised a valid point. Although an empty document in Firestore will *still* count towards the `data` array's length (to my understanding), I agree that it could potentially cause the "Your shopping list is empty" message not to display. To address this, *maybe* you could create a helper function in `List.jsx` like this:

// we're checking if *every* document in the data array is **empty**
```
const isListTrulyEmpty = (data) => {
    return data.length === 0 || data.every(item => Object.keys(item).length === 0);
}
```

Then:

```
{isListTrulyEmpty(data) ? (
	<p>Your shopping list is empty. Click on "Add item" to begin!</p>
) : (
    blah blah
)}
```

This approach checks both the array's length and each document's content for a more accurate conditional render.

Please let me know your thoughts on this.

Comment (rdpfeifle) [reply src/api/firebase.js:1]: (Additionally, even if we encounter issues down the road, I'm confident that we can all come together and figure it out ✨💃🏻)

Comment (CorinaMurg) [reply src/api/firebase.js:1]: We wanted to try to implement the optional feature (ie adding a new list to Firestore as soon as it is created so users can share the token for an empty list) but as far as we can tell, Firestore does not allow a new collection unless it contains "something." That's why the empty object. We did take care of filtering out that empty object by using only objects that have a name property. Having said that, we worked on two different versions, partly because we could not anticipate if there were other drawbacks to having that empty object idling in the background 🤔

Comment (CorinaMurg) [reply src/api/firebase.js:1]: But if we decide to keep it, I believe the filter we use right at the top of the `List` component should take care of everything: `const realData = data.filter((item) => item.name)`. So as long as we reference the `realData` throughout the `List` we should have no problems. (Well, let's hope 😄)

File: src/views/Home.jsx
Comment (cafloyd) [inline src/views/Home.jsx:45]: **Praise:** using `setTimeout` to make the message disappear after a few seconds was a nice touch here!

File: src/views/Home.jsx
Comment (cafloyd) [inline src/views/Home.jsx:71]: **Praise:** Love this attention to detail! Great call paying attention to keyboard focus!

File: src/views/List.jsx
Comment (cafloyd) [inline src/views/List.jsx]: **Nitpick:** We probably want to remove this `console.log` here!

File: src/views/List.jsx
Comment (cafloyd) [inline src/views/List.jsx]: **Suggestion:** I know this is just what was already in the component, but I think this was intended as placeholder text until there was content on the page. We can probably remove it, unless you all like it there!

**Review by cafloyd — COMMENTED** [review]
Fantastic work, team!! I love all the helpful comments you left and the really useful PR description!

**Nitpick:** I left a couple suggestions for code clean up - let's remove at least the `console.log` before merging this PR. I'll approve once that's done!

**Suggestion (non-blocking):** A [suggestion I made](https://github.com/the-collab-lab/tcl-63-smart-shopping-list/pull/21#discussion_r1311054365) on the other pair's PR might also be useful here. This is completely optional and not at all a blocker for this issue, but if you want, you can consider sanitizing any data that is being input by the user before you use it. See my other comment for additional context, but if the team wants to move forward with this, it might make the most sense to create a helper function that can be reused wherever there is a text input.

**Review by cafloyd — APPROVED** [review]
Thanks for cleaning up the code! Approving the PR, up to you all if you want to implement any data sanitizing - the other team implemented it in [this commit](https://github.com/the-collab-lab/tcl-63-smart-shopping-list/pull/21/commits/e62c26016b90b533160fffabe7b6398972df11a8)

File: src/views/List.jsx
Comment (CorinaMurg) [reply src/views/List.jsx]: @cafloyd  Thank you so much Caitlin! Should we also remove OPTION A (code and comments) re: `createNewList` function? We weren't sure if OPTION B version would pass the "good code" test since it relies on adding an empty object (which we then filter out in the `List`). It made sense to write the function that way because it reflects Firestore's behavior when operated manually (when adding a collection manually it can't be saved unless it contains an item) but we wanted to have a mentor' approval first. 😊

Comment (CorinaMurg) [reply src/views/List.jsx]: @cafloyd  Regarding the sanitizing code, since the other PR was already merged, it should not matter if we include the regex on our end as well, no? I'm assuming it will only create an extra conflict. (Speaking of conflicts, I think I finally got my wish. THREE conflicts! Many thanks to you capricious gods of coding!  😂)

Comment (cafloyd) [reply src/views/List.jsx]: @CorinaMurg - 

1) I thought you all might be leaving those 'Option A' tips in the comments in case the issues Aslan described became an issue. If that's the reason, I don't mind leaving the comments there to make things easier for your future selves. We can remove them if we find it's not an issue.

As for whether this passes the 'good code' test, I don't personally have a problem with it. This approach was mentioned as an acceptable one in the ticket. If it breaks something, then we can revisit 😅 

2) The other team added sanitizing for the filter input, but we could also add it for the token input. If you wanted to add it to the token input, there are three options I see:

	a) Add the regex to the token input, and possibly DRY up the code in a future PR by pulling the logic out to a helper function.
	b) Pull the other pair's code into your branch to address merge conflicts. Once they've been fixed, pull the regex logic out to a helper function and use it for the token input too.
	c) Merge your PR as-is, and create a follow-up PR to create the helper function and add it to the token input.

Since you already plan on working through merge conflicts, maybe Option B makes sense here. You can always start working through the merge conflicts, and if you're not able to get to the sanitizing you can leave it for the future!

Will leave this up to you and @rdpfeifle ! Good luck with the merge conflicts 🎉 😆  and ping me on Slack if you run into any issues!!

Comment (CorinaMurg) [reply src/views/List.jsx]: All done! Thank you for the thorough explanations, they were super helpful!

**Pull request merged**
